### PR TITLE
Enhancement/17/peer reputation

### PR DESCRIPTION
**Pull request by Prodicode**
#17 

* Added reputation.ex API
* Sets/Loads reputation on peer connect
* Removes reputation on invalid block, garbage data, corrupt data and not requested peer connection

### CODE REVIEW

File: apps/purple/lib/purple/workers/network/tcp_worker.ex [hunk 1]
```diff
@@ -129,13 +131,21 @@ defmodule Purple.Workers.Network.TcpWorker do
...
+          #-2 reputation
+          {:ok, {ip, port}} = :inet.peername(sock)
+          :ok = Reputation.remove_node_rep(:inet.ntoa(ip), 2)
           {:noreply, state}
       end
     else
       {:error, :badcrc} ->
         # TODO: Handle data corruption case
+
+        #-2 reputation
```
Comment (octavonce) [inline apps/purple/lib/purple/workers/network/tcp_worker.ex]: This is the case when the packet didn't get sent properly. We shouldn't decrease the reputation for this.

File: apps/purple/lib/purple/network/reputation.ex [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    a node's reputation.
+  """
+
+  require Logger
+
+  @doc """
+    Sets reputation for a node ip
+  """
+  @spec set_node_rep(binary, any) :: :ok
+  def set_node_rep(ip, val) do
```
Comment (octavonce) [inline apps/purple/lib/purple/network/reputation.ex]: Please add function guards.

File: apps/purple/lib/purple/network/reputation.ex [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        :ok = NodeStorage.put("reputation.#{ip}", val)
+      {:ok, term} ->
+        :ok = NodeStorage.delete("reputation.#{ip}")
+        :ok = NodeStorage.put("reputation.#{ip}", val)
+    end
+    :ok
+  end
+
+  @doc """
+  Removes reputation from a node
```
Comment (octavonce) [inline apps/purple/lib/purple/network/reputation.ex]: Please be consistent with the style of the other code.

This should be like this:
```elixir
@doc """
  Removes reputation from a node
"""
```

File: apps/purple/lib/purple/network/reputation.ex [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        :error
+    end
+  end
+
+  @doc """
+  Deltes a node's reputation
+  """
+
+  @spec delete_node_rep(binary) :: :ok | :error
+  def delete_node_rep(ip) do
```
Comment (octavonce) [inline apps/purple/lib/purple/network/reputation.ex]: Please add function guards and property tests.

File: apps/purple/lib/purple/network/pipeline/packet_handler.ex [hunk 5]
```diff
@@ -48,36 +51,55 @@ defmodule Purple.Network.PacketHandler do
...
-        
+
         peers
         |> Enum.filter(fn %{peer_id: peer_id} -> peer_id != node_id end)
-        |> Enum.each(fn %{ip: ip, peer_id: peer_id} -> 
+        |> Enum.each(fn %{ip: ip, peer_id: peer_id} ->
           if not Peers.reached_peer_limit? do
-            Peers.connect(peer_id, ip, @port) 
+            Peers.connect(peer_id, ip, @port)
+            #Set reputation
```
Comment (octavonce) [inline apps/purple/lib/purple/network/pipeline/packet_handler.ex]: Please add a single space after the '#'.

This should be like this:
```elixir
# Set reputation
```

File: apps/purple/lib/purple/network/pipeline/packet_handler.ex [hunk 6]
```diff
@@ -48,36 +51,55 @@ defmodule Purple.Network.PacketHandler do
...
-      {:error, _} -> 
+      {:error, _} ->
         :ok
-      _ -> 
-        # TODO: Handle case when we haven't requested peers 
+      _ ->
+        # TODO: Handle case when we haven't requested peers
         # or the request has expired.
+        {:ok, node_id} = NodeStorage.get("node_id")
+        peers
```
Comment (octavonce) [inline apps/purple/lib/purple/network/pipeline/packet_handler.ex]: You are decreasing the reputation of every peer here.

Just call remove_node_rep directly:
```elixir
ip
|> :inet.ntoa()
|> Reputation.remove_node_rep(5)
```

**Comment by Prodicode** [comment]
The requested changes were addressed.

**Comment by Prodicode** [comment]
Finished in 15.0 seconds
48 tests, 0 failures

**Comment by Prodicode** [comment]
Added reputation property test:

Finished in 14.8 seconds
49 tests, 0 failures

**Pull request merged**